File: src/main/java/com/leijendary/spring/iamtemplate/generator/SnowflakeIdGenerator.java
package com.leijendary.spring.iamtemplate.generator;

import com.leijendary.spring.iamtemplate.worker.SnowflakeIdWorker;
import org.hibernate.HibernateException;
import org.hibernate.engine.spi.SharedSessionContractImplementor;
import org.hibernate.id.IdentifierGenerator;

import java.io.Serializable;

public class SnowflakeIdGenerator implements IdentifierGenerator {

    public static final String STRATEGY = "com.leijendary.spring.iamtemplate.generator.SnowflakeIdGenerator";

    private static final SnowflakeIdWorker WORKER = new SnowflakeIdWorker();

    @Override
    public Serializable generate(final SharedSessionContractImplementor session,
                                 final Object object) throws HibernateException {
        return WORKER.nextId();
    }
}

File: src/main/kotlin/com/leijendary/spring/template/iam/core/generator/SnowflakeIdGenerator.kt
package com.leijendary.spring.template.iam.core.generator

import com.leijendary.spring.template.iam.core.worker.SnowflakeIdWorker
import org.hibernate.engine.spi.SharedSessionContractImplementor
import org.hibernate.id.IdentifierGenerator
import java.io.Serializable

class SnowflakeIdGenerator : IdentifierGenerator {
    companion object {
        const val STRATEGY = "com.leijendary.spring.template.iam.core.generator.SnowflakeIdGenerator"
        private val WORKER = SnowflakeIdWorker()
    }

    override fun generate(session: SharedSessionContractImplementor, any: Any): Serializable = WORKER.nextId()
}
